File: cplus/leetcode/202603/algo3348.cpp
#include <bits/stdc++.h>
using namespace std;

// 三、字符串贪心 - 3.1 字典序最小/最大 倒序贪心
class Solution {
public:
    string smallestNumber(string num, long long t) {
        // 参考题解 数位 DP
        vector<int> primes = {2, 3, 5 , 7}; auto tmp = t;
        int cnt = 0;

        for (auto &p : primes) {
            while (tmp % p == 0) {
                tmp /= p; cnt ++;
            }
        }
        if (tmp > 1) return "-1";

        cnt = max(1, cnt - (int)num.length() + 1);

        string s(cnt, '0');
        s.insert(s.end(), num.begin(), num.end());

        int n = s.length();
        string ans(n, '0');

        unordered_map<long long, bool> memo;
        auto dfs = [&](this auto&& dfs, int i, long long t, bool isLimit) -> bool {
            if (i == n) return t == 1;
            long long mask = (long long) i << 50 | (long long)isLimit << 48 | t;
            if (memo.count(mask)) return memo[mask];

            if (isLimit && i < cnt && dfs(i+1, t, isLimit)) return memo[mask] = true;

            int low = isLimit?s[i] - '0': 0;
            for (int d=max(low, 1); d<10; d++) {
                ans[i] = d + '0';
                if (dfs(i+1, t/gcd(t,d), isLimit && d == low)) return memo[mask] = true;
            }
            ;
            return memo[mask] = false;
        };

        dfs(0, t, true);
        int inx = 0; while (ans[inx] == '0') inx ++;
        return ans.substr(inx);

    }    
};

File: python/algo/202606/algo3348.py
from functools import cache
from math import gcd


class Solution:
    def smallestNumber(self, num: str, t: int) -> str:
        # 参考题解 数位DP，有讲解豁然开朗
        cnt, tmp, primes = 0, t, [2,3,5,7]
        for p in primes:
            while tmp % p == 0:
                tmp /= p
                cnt += 1
        
        if tmp > 1: return "-1"

        cnt = max(cnt - len(num) + 1, 1)
        s = "0" * cnt + num
        n = len(s)
        ans = ['0'] * n

        @cache
        def dfs(i: int, t:int, isLimit: bool) -> bool:
            if i == n: return t == 1

            if isLimit and i < cnt and dfs(i+1, t, isLimit): return True

            low = int(s[i]) if isLimit else 0
            for d in range(max(low, 1), 10):
                ans[i] = str(d)
                if dfs(i+1, t // gcd(t, d), isLimit and low == d): return True
            return False

        dfs(0, t, True)
        dfs.cache_clear()
        return "".join(ans).lstrip("0") 